File: webapp/src/lib/authenticity.ts
// Shared job-authenticity ledger — community real/dubious/scam feedback.
import { api } from "./api";

export interface Authenticity {
  key: string;
  company: string;
  title: string;
  verdict: string;
  employer_status: string;
  employer_detail: string;
  min_authenticity_score: number;
  tally: { legit: number; dubious: number; scam: number };
  reasons: string[];
  your_vote: string | null;
  last_checked_at: string | null;
}

export const VERDICT_META: Record<string, { label: string; icon: string; cls: string }> = {
  verified_real: { label: "Verified real", icon: "ri-verified-badge-fill", cls: "bg-primary-100 text-primary-800" },
  likely_real: { label: "Likely real", icon: "ri-checkbox-circle-line", cls: "bg-primary-50 text-primary-700" },
  unverified: { label: "Unverified", icon: "ri-question-line", cls: "bg-background-200 text-foreground-600" },
  dubious: { label: "Dubious", icon: "ri-error-warning-line", cls: "bg-secondary-100 text-secondary-900" },
  likely_scam: { label: "Likely scam", icon: "ri-alarm-warning-fill", cls: "bg-accent-100 text-accent-900" },
};

export const EMPLOYER_META: Record<string, string> = {
  listed: "Listed on the employer site",
  not_found: "Not found on the employer site",
  invalid_domain: "No verifiable employer domain",
  unknown: "Employer site not checked",
};

export function getJobAuthenticity(jobId: string) {
  return api<Authenticity>(`/api/v1/authenticity/job/${jobId}`);
}

export function reportJob(jobId: string, verdict: "legit" | "dubious" | "scam", reason = "") {
  return api<Authenticity>(`/api/v1/authenticity/job/${jobId}/report`, { method: "POST", body: { verdict, reason } });
}

export function verifyEmployer(jobId: string) {
  return api<Authenticity>(`/api/v1/authenticity/job/${jobId}/verify-employer`, { method: "POST", body: {} });
}

export function listFlagged() {
  return api<Authenticity[]>("/api/v1/authenticity/flagged");
}

File: src/jobsearch/api/routers/authenticity.py
"""Shared job-authenticity ledger — community real/dubious/scam feedback."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, status

from jobsearch.api.deps import CurrentUser, StateDep
from jobsearch.api.schemas import AuthenticityOut, JobReportRequest
from jobsearch.models import JobAuthenticityRecord, JobPosting, ReportVerdict

router = APIRouter(prefix="/api/v1/authenticity", tags=["authenticity"])


def _out(rec: JobAuthenticityRecord, user_id: str) -> AuthenticityOut:
    return AuthenticityOut(
        key=rec.key,
        company=rec.company,
        title=rec.title,
        verdict=rec.verdict.value,
        employer_status=rec.employer_status.value,
        employer_detail=rec.employer_detail,
        min_authenticity_score=rec.min_authenticity_score,
        tally=rec.tally(),
        reasons=rec.reasons,
        your_vote=rec.votes.get(user_id),
        last_checked_at=rec.last_checked_at.isoformat() if rec.last_checked_at else None,
    )


def _job(job_id: str, state: StateDep) -> JobPosting:
    job = state.jobs.get(job_id)
    if job is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "job not found")
    return job


def _score(job: JobPosting, state: StateDep) -> int:
    v = state.verifications.get(job.id) or state.verification.verify(job)
    state.verifications[job.id] = v
    return v.authenticity_score


@router.get("/job/{job_id}", response_model=AuthenticityOut)
def job_authenticity(job_id: str, user: CurrentUser, state: StateDep) -> AuthenticityOut:
    """The shared verdict for a job (community reports + automated signals)."""
    job = _job(job_id, state)
    rec = state.authenticity.snapshot(job, authenticity_score=_score(job, state))
    return _out(rec, user.id)


@router.post("/job/{job_id}/report", response_model=AuthenticityOut)
def report_job(job_id: str, body: JobReportRequest, user: CurrentUser, state: StateDep) -> AuthenticityOut:
    """Report a posting as legit / dubious / scam (one vote per user)."""
    job = _job(job_id, state)
    try:
        verdict = ReportVerdict(body.verdict)
    except ValueError as exc:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, f"unknown verdict '{body.verdict}'") from exc
    rec = state.authenticity.report(
        user.id, job, verdict, reason=body.reason, authenticity_score=_score(job, state)
    )
    return _out(rec, user.id)


@router.post("/job/{job_id}/verify-employer", response_model=AuthenticityOut)
def verify_employer(job_id: str, user: CurrentUser, state: StateDep) -> AuthenticityOut:
    """Check the employer's site to confirm the posting is really available."""
    job = _job(job_id, state)
    rec = state.authenticity.check_employer(job, authenticity_score=_score(job, state))
    return _out(rec, user.id)


@router.get("/flagged", response_model=list[AuthenticityOut])
def flagged(user: CurrentUser, state: StateDep, limit: int = Query(50, ge=1, le=200)) -> list[AuthenticityOut]:
    """The shared scam-watch list — postings the community flagged as dubious/scam."""
    return [_out(r, user.id) for r in state.authenticity.flagged(limit=limit)]
